### PR TITLE
Refactor C++ `active_list_type` and remove Python `ActiveList`

### LINKED ISSUE #2485: Replace active_list with std::vector
We should be able to replace active_list_type with a std:vector by changing the construction to getting all the incidences instead of the `mode` property. That should enable us to remove quite a few code paths. Should check why this was implemented initially.

### LINKED ISSUE #2745: Replace CWrap with "extender functions" in `res.enkf.ActiveList` (as a first)
I've discovered that pybind11 allows extending classes defined via C++ in Python. That is:

```c++
/* C++ */
struct Foo {
    int val = 0;
};

PYBIND11_MODULE(_lib, m) {
    py::class_<Foo>(m, "Foo")
        .def(py::init<>())
        .def_readwrite("val", &Foo::val);

    m.def("get_instance", [] {
        return new Foo;  // Create a Foo object from C++ rather than Python
    });
}
```

```py
# Python
from res._lib import Foo, get_instance

# Before adding a new class instance function
pre = get_instance()
pre.val = 5
print(pre.val) #=> 5

# Add instance function
def test(self):
    return self.val + 10
Foo.test = test

# After adding a new class instance function
print(pre.test()) #=> 15
print(Foo().test()) #=> 10
print(get_instance().test()) #=> 10
```

---

This solves a major problem I've had with CWrap, which is that one has to define two separate classes -- one C implementation and one Python wrapper class. Now there is a single class defined in C++, while retaining the ability to write Python code for additional functionality.

The wrapper class does not behave like a normal Python class, since its `__init__` will not necessarily be called, any field variables that were defined in Python will disappear after a roundtrip via C, and requires that C's manual object lifetime management bleed into Python (via `alloc` and `free` functions).

On the other side, C++ has to use opaque pointers when passing objects around, rather than references and such as one would expect from modern C++. In practice we would like for objects to not be allocated with `malloc`/`new`, but be allocated on the stack and moved around. This extension function proposal would allow it.

Finally, this removes the need for registering CWrap classes in `REGISTERED_TYPES` and the use of `BaseCClass`, which will reduce the amount of `cwrap` usage in the code base.

---

To make the code cleaner, I would create a decorator function `@extend`, taking a class, and adding the decorated function to the class, as above. Supporting properties and `@staticmethod`/`@classmethod` should also be investigated.

```py
from res._lib.enkf import ActiveList

__all__ = ["ActiveList"]

@extend(ActiveList)
def __repr__(self):
    ...

### PR DESCRIPTION
**Pull request by pinkwah**
I went into this expecting to be able to do #2745 . However, it turns out that `ActiveList` isn't actually used in Python, except for the tests. The class also gives me vibes of unfinishedness. If it's meant to be used in Python, perhaps we should go back to the drawing board.

As such, this PR removes the Python class entirely and refactors the C code into okayish C++.

To deal with test usage I've created a new submodule `res._lib.test_support` (ie, supports for testing). I reckon we can use this location for other place where the CWrap overexposes the code that is meant for only for tests.

**Issue**
Resolves #2485

### CODE REVIEW

File: libres/lib/include/ert/enkf/active_list.hpp [hunk 1]
```diff
@@ -1,52 +1,31 @@
...
-void active_list_copy(active_list_type *target, const active_list_type *src);
-
-UTIL_IS_INSTANCE_HEADER(active_list);
-
-#ifdef __cplusplus
+#pragma once
+
+#include <unordered_set>
+
+namespace enkf {
```
Comment (joakim-hove) [inline libres/lib/include/ert/enkf/active_list.hpp:5]: When switching to use C++ namespace I would recommend using something else than `enkf` - ERT has for a long time been unable to execute the EnKF algorithm. Do not have an obvious suggestion, but this is part of the configuration of localization in updates - so something with `update` and/or `local`?

Comment (pinkwah) [reply libres/lib/include/ert/enkf/active_list.hpp:5]: I was using the name of the directory as the namespace. Perhaps we should not have an `enkf` directory at all and move this particular code somewhere else?

Comment (joakim-hove) [reply libres/lib/include/ert/enkf/active_list.hpp:5]: My suggestion would be to move it to the `analysis/` directory alongside the `update.cpp` file and put it in the `analysis::` namespace.

**Comment by pinkwah** [comment]
Blocked due to this causing a breaking change in `semeio` and there's a cyclical dependency which I am not sure how to resolve.

**Comment by oyvindeide** [comment]
It _might_ be possible to replace it with an index list, as that is what it essentially is, with some additional features, some additional discussion here: https://github.com/equinor/ert/issues/2485

**Comment by pinkwah** [comment]
After having discussions, I think we should drop the specialised `ActiveList` structure entirely and bring this functionality directly into the observation types.

**Comment by joakim-hove** [comment]
> After having discussions, I think we should drop the specialised ActiveList structure entirely and bring this functionality directly into the observation types.

I understand the desire to get rid of a "class" which is essentially just a `std::vector<int>`; the main reason for it to exist in it's current form is because of the size concept. The available information through the lifetime goes like this:

1. At creation we do not know the size - but it is by default created with the mode `ALL_ACTIVE`
2. When we actually need it - we know from calling scope what the `ALL_ACTIVE` default size.

The metadata aorund this is is handled in the current ActiveList. My suggestion would be to do nothing while there are ongoing changes to analysis+++ - then when that dust has settled it maybe falls out very naturally how to proceed.

**Comment by pinkwah** [comment]
The idea is to move this behaviour up a layer and let the end-users work with an `std::vector` (or a numpy object in Python) directly without an additional layer of indirection. You can achieve this with a `std::optional<AppropriateContainer<int>>` (`std::vector` is not an appropriate container here):

- If `std::nullopt`, assume `ALL_ACTIVE`
- Otherwise it's `PARTLY_ACTIVE`
- Let the user:
  1. Get the list with `::get_or_create_active_list()`
  2. Delete the list with `::remove_active_list()`

However, this is a breaking change in the user-facing part of ERT so it shouldn't be high on our priority lists.

**Pull request closed without merging**